client/src/main/scala/com.gu.contentapi.client/BackoffStrategy.scala

```diff
@@ -85,17 +76,35 @@ object ContentApiBackoff {
   private def constant(
     min: Duration = Duration(defaultMinimumInterval, TimeUnit.MILLISECONDS),
     maxAttempts: Int = defaultMaxAttempts
-  )(implicit executor: ScheduledExecutor): Constant = {
+  ): Constant = {
     val ln = Math.max(min.toMillis, defaultMinimumInterval)
     val mx = if (maxAttempts > 0) maxAttempts else 1
     Constant(Duration(ln, TimeUnit.MILLISECONDS), 0, mx)
   }
+}
 
-  private def attempt(backoff: ContentApiBackoff, retry: Int, operation: Int => Future[HttpResponse])(implicit context: ExecutionContext): Future[HttpResponse] = {
-    operation(retry)
-      .recoverWith {
-        case _: ContentApiRecoverableException => backoff.execute(operation)
-      }
+object Retry {
+  type RetryAttempt = Int
+  def withRetry[A](backoffStrategy: BackoffStrategy, retryPredicate: A => Boolean)(operation: RetryAttempt => Future[A])(implicit executor: ScheduledExecutor, ec: ExecutionContext): Future[A] = {
+    def loop(backoffStrategy: BackoffStrategy): Future[A] = backoffStrategy match {
+      case r: Retryable => operation(r.attempts).flatMap {
+          case result if retryPredicate(result) => executor.sleepFor(r.delay).flatMap(_ => loop(backoffStrategy.increment))
+          case result => Future.successful(result)
+        }
+      case RetryFailed(attempts) =>
+        Future.failed(ContentApiBackoffException(s"Backoff failed after $attempts attempts"))
+    }
+    loop(backoffStrategy)
   }
-
 }
+
+object HttpRetry {
+
+  private def canRetry(implicit executionContext: ExecutionContext): HttpResponse => Boolean = {
+    case IsRecoverableHttpResponse() => true
```

Comment:
My preference would be to use a 'regular' function; this syntax could just confuse people not familiar with pattern matching that `unapply()` provides.

Reply:
Done. 
I agree with you that in this case, simple methods will suffice.

Not agreeing on the confusion argument though as I expect Scala devs to know (or learn) how pattern matching works under the hood.